pkg/pool-weighted/contracts/managed/ManagedPoolCircuitBreakerLib.sol

```diff
@@ -0,0 +1,248 @@
+
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../lib/ValueCompression.sol";
+
+/**
+ * @title Managed Pool Circuit Breaker Library
+ * @notice Library for storing and manipulating state related to circuit breakers in ManagedPool.
+ * @dev BPT price = supply / balance / weight. The lower and upper bounds are percentages.
+ * For instance, a lower bound of 0.8 means the circuit breaker will trip if the token price
+ * drops more than 20% (relative to other tokens). An upper bound ratio of 2.0 means it will trip
+ * if the price more than doubles.
+ */
+library ManagedPoolCircuitBreakerLib {
+    using ValueCompression for uint256;
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+
+    struct CircuitBreakerParams {
+        uint256 referenceBptPrice;
+        uint256 currentWeightFactor;
+        uint256 lowerBoundPercentage;
+        uint256 upperBoundPercentage;
+    }
+
+    // Store circuit breaker information per token
+    // [  24 bits |  24 bits |   16 bits   |   16 bits   |    64 bits    |  112 bits |
+    // [ UB ratio | LB ratio | upper bound | lower bound | weight factor | ref price |
+    // |MSB                                                                       LSB|
+    uint256 private constant _REFERENCE_BPT_PRICE_OFFSET = 0;
+    uint256 private constant _WEIGHT_FACTOR_OFFSET = _REFERENCE_BPT_PRICE_OFFSET + _BPT_PRICE_WIDTH;
+    uint256 private constant _LOWER_BOUND_PCT_OFFSET = _WEIGHT_FACTOR_OFFSET + _WEIGHT_FACTOR_WIDTH;
+    uint256 private constant _UPPER_BOUND_PCT_OFFSET = _LOWER_BOUND_PCT_OFFSET + _BOUND_PERCENTAGE_WIDTH;
+    uint256 private constant _LOWER_BOUND_RATIO_CACHE_OFFSET = _UPPER_BOUND_PCT_OFFSET + _BOUND_PERCENTAGE_WIDTH;
+    uint256 private constant _UPPER_BOUND_RATIO_CACHE_OFFSET = _LOWER_BOUND_RATIO_CACHE_OFFSET +
+        _BOUND_RATIO_CACHE_WIDTH;
+
+    uint256 private constant _BOUND_PERCENTAGE_WIDTH = 16;
+    uint256 private constant _BOUND_RATIO_CACHE_WIDTH = 24;
+    uint256 private constant _WEIGHT_FACTOR_WIDTH = 64;
+    uint256 private constant _BPT_PRICE_WIDTH = 112;
+
+    // We compress the ratios into 16 bits from a range of [0, 10e18], chosen to allow the upper bound to exceed 1.
+    // For consistency, use the same maximum uncompressed value, even though we the lower bound is less than 1.
+    uint256 private constant _MAX_BOUND_PERCENTAGE = 10e18; // FP 10
+
+    /**
+     * @notice Returns the reference BPT price, reference weight factor, and upper and lower bounds for a given token.
+     * @dev If a bound value is zero, it means there is no circuit breaker in that direction for the given token.
+     * @param circuitBreakerState - The bytes32 state of the token of interest.
+     */
+    function getCircuitBreakerFields(bytes32 circuitBreakerState)
+        internal
+        pure
+        returns (
+            uint256,
+            uint256,
+            uint256,
+            uint256
+        )
```

Comment:
I'll return the struct, that will address both issues. I'm using a struct both to help with the stack, and in case we need to change the fields later.

Reply:
If using structs keep https://github.com/ethereum/solidity/issues/13475 in mind